src/pudl/analysis/ferc1_eia_train.py

```diff
@@ -0,0 +1,720 @@
+
+    Returns:
+        pd.DataFrame: The validated FERC-EIA dataframe you're trying to add to the
+            training data.
+    """
+    logger.info("Validating overrides")
+    # When there are NA values in the verified column in the excel doc, it seems that
+    # the TRUE values become 1.0 and the column becomes a type float. Let's replace
+    # those here and make it a boolean.
+    validated_connections["verified"] = validated_connections["verified"].replace(
+        {1: True, np.nan: False}
+    )
+    # Make sure the verified column doesn't contain non-boolean outliers. This will fail
+    # if there are bad values.
+    validated_connections.astype({"verified": pd.BooleanDtype()})
+
+    # From validated records, get only records with an override
+    only_overrides = (
+        validated_connections[validated_connections["verified"]]
+        .dropna(subset=["record_id_eia_override_1"])
+        .reset_index()
+        .copy()
+    )
+
+    # Make sure that the override EIA ids actually match those in the original FERC-EIA
+    # record linkage.
+    actual_eia_ids = ppl.record_id_eia.unique()
+    _check_id_consistency(
+        "record_id_eia_override_1",
+        only_overrides,
+        actual_eia_ids,
+        "values that don't exist",
+    )
+
+    # It's unlikely that this changed, but check FERC id too just in case!
+    actual_ferc_ids = ferc1_eia.record_id_ferc1.unique()
+    _check_id_consistency(
+        "record_id_ferc1", only_overrides, actual_ferc_ids, "values that don't exist"
+    )
+
+    # Make sure there are no duplicate EIA id overrides
+    logger.debug("Checking for duplicate override ids")
+    assert (
+        len(
+            override_dups := only_overrides[
+                only_overrides["record_id_eia_override_1"].duplicated(keep=False)
+            ]
+        )
+        == 0
+    ), f"Found record_id_eia_override_1 duplicates: \
+    {override_dups.record_id_eia_override_1.unique()}"
+
+    if not allow_mismatched_utilities:
+        # Make sure the EIA utility id from the override matches the PUDL id from the FERC
+        # record. Start by mapping utility_id_eia from PPL onto each
+        # record_id_eia_override_1.
+        logger.debug("Checking for mismatched utility ids")
+        only_overrides = only_overrides.merge(
+            ppl[["record_id_eia", "utility_id_eia"]].drop_duplicates(),
+            left_on="record_id_eia_override_1",
+            right_on="record_id_eia",
+            how="left",
+            suffixes=("", "_ppl"),
+        )
+        # Now merge the utility_id_pudl from EIA in so that you can compare it with the
+        # utility_id_pudl from FERC that's already in the overrides
+        only_overrides = only_overrides.merge(
+            utils_eia860[["utility_id_eia", "utility_id_pudl"]].drop_duplicates(),
+            left_on="utility_id_eia_ppl",
+            right_on="utility_id_eia",
+            how="left",
+            suffixes=("", "_utils"),
+        )
```

Comment:
if this is the only use of this `utils_eia860`, why not just get this info from the `ppe`? (trying to reduce the number of dfs flowing into your tests)

Reply:
That's a good idea